src/Shared/Targets/Wrappers/AspNetBufferingTargetWrapper.cs

```diff
@@ -168,48 +202,65 @@ protected override void CloseTarget()
         /// <param name="logEvent">The log event.</param>
         protected override void Write(AsyncLogEventInfo logEvent)
         {
-            var buffer = GetRequestBuffer();
+            var buffer = GetOrCreateRequestBuffer(HttpContextAccessor.HttpContext);
             if (buffer != null)
             {
-                WrappedTarget.PrecalculateVolatileLayouts(logEvent.LogEvent);
-
-                buffer.Append(logEvent);
+                WrappedTarget?.PrecalculateVolatileLayouts(logEvent.LogEvent);
                 InternalLogger.Trace("Appending log event {0} to ASP.NET request buffer.", logEvent.LogEvent.SequenceID);
+                buffer.Append(logEvent);
             }
             else
             {
                 InternalLogger.Trace("ASP.NET request buffer does not exist. Passing to wrapped target.");
-                WrappedTarget.WriteAsyncLogEvent(logEvent);
+                WrappedTarget?.WriteAsyncLogEvent(logEvent);
             }
         }
 
-        private NLog.Web.Internal.LogEventInfoBuffer GetRequestBuffer()
+        /// <summary>
+        ///
+        /// </summary>
+        /// <param name="context"></param>
+        /// <returns></returns>
+        private Internal.LogEventInfoBuffer GetOrCreateRequestBuffer(
+#if ASP_NET_CORE
+            HttpContext context)
+#else
+            HttpContextBase context)
+#endif
         {
-            HttpContext context = HttpContext.Current;
-            if (context == null)
+            if (context != null && context.Items[DataSlot] == null)
             {
-                return null;
+                context.Items[DataSlot] =
+                    new Internal.LogEventInfoBuffer(BufferSize, GrowBufferAsNeeded, BufferGrowLimit);
             }
-
-            return context.Items[dataSlot] as NLog.Web.Internal.LogEventInfoBuffer;
-        }
-
-        private void OnBeginRequest(object sender, EventArgs args)
-        {
-            InternalLogger.Trace("Setting up ASP.NET request buffer.");
-            HttpContext context = HttpContext.Current;
-            context.Items[dataSlot] = new NLog.Web.Internal.LogEventInfoBuffer(BufferSize, GrowBufferAsNeeded, BufferGrowLimit);
+            return context?.Items?[DataSlot] as Internal.LogEventInfoBuffer;
         }
 
+        /// <summary>
+        ///
+        /// </summary>
+        /// <param name="sender"></param>
+        /// <param name="args"></param>
         private void OnEndRequest(object sender, EventArgs args)
         {
-            var buffer = GetRequestBuffer();
-            if (buffer != null)
+            var context = HttpContextAccessor.HttpContext;
```

Comment:
Think this logic should be in the Middleware / HttpModule. So the it is the Middleware / HttpModule that performs the lookup in the active HttpContext for any active buffer and "commits" the buffered logevents.

The buffer should have an owner-reference to the `AspNetBufferingTargetWrapper`-target that created the buffer, so the Middleware / HttpModule can notify the owner that it is time to write.

Thus removing `EndRequest`-EventHandler.

Reply:
But I guess that means the `DataSlot` must be static, and we should be able to handle a linked-list of buffers, when having multiple AspNetBufferingTargetWrapper-objects.

So the buffer should both have an "owner" but also a "next"-reference to allow having multiple `AspNetBufferingTargetWrapper`-target register multiple buffers for the same HttpContext-Item-Dataslot. And then the Middleware / HttpModule can notify the entire chain of buffers that HttpRequest has completed, and it is time to write.

Reply:
We could do that certainly.  Maybe there is a way to keep DataSlot not static and still have the middleware/http module do the buffered log events.  
I see LoggingConfiguration has an AllTargets ReadOnlyCollection.  Could the middleware/ http module loop thru that collection and if the target is a AspNetBufferingTargetWrapper, call an internal method on the target that would commit the LogEvents in that target's own private DataSlot?  Could add an extension method to LoggingConfiguration called GetAllTargetsOfType<T>() for example as well.

Reply:
Committed a change showing this approach.

Reply:
I don't like global static variables, which is why I'm suggesting using the HttpContext as the transport for handling the request-buffering.

Replacing the static-event-handler with the global static LogManager is not a good trade. Also there is a performance penalty in LogFactory.Configuration-property, as it contains a monitor-lock.

Reply:
Those are very valid points.  I have removed LogManager and instead use a Dictionary, this should be simialr to your request now.

Reply:
The only static is an object used as a key to lookup the Dictionary in the HttpContext.Items.